subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -177,6 +197,46 @@ public boolean test(String element) {
         }
     }
 
+    private interface SymbolicLinkMappings {
+        String remapAbsolutePath(Path path);
+        @CheckReturnValue
+        SymbolicLinkMappings withNewMapping(String source, String target, RelativePathTracker currentRelativePathTracker);
+        Iterable<String> getRemappedSegments(Iterable<String> segments);
+    }
+
+    private static class DefaultSymbolicLinkMappings implements SymbolicLinkMappings {
+        private final ImmutableList<SymbolicLinkMapping> symbolicLinkMappings;
+        private final Iterable<String> prefixRelativePath;
+
+        public DefaultSymbolicLinkMappings(ImmutableList<SymbolicLinkMapping> symbolicLinkMappings, Iterable<String> prefixRelativePath) {
+            this.symbolicLinkMappings = symbolicLinkMappings;
+            this.prefixRelativePath = prefixRelativePath;
+        }
+
+        @Override
+        public String remapAbsolutePath(Path dir) {
+            String targetAbsolutePath = dir.toString();
+            for (SymbolicLinkMapping symbolicLinkMapping : symbolicLinkMappings) {
+                targetAbsolutePath = symbolicLinkMapping.remapPath(targetAbsolutePath);
+            }
+            return targetAbsolutePath;
+        }
+
+        @Override
+        public SymbolicLinkMappings withNewMapping(String source, String target, RelativePathTracker currentRelativePathTracker) {
+            ImmutableList<SymbolicLinkMapping> newMappings = ImmutableList.<SymbolicLinkMapping>builderWithExpectedSize(symbolicLinkMappings.size() + 1)
+                .add(new SymbolicLinkMapping(source, target))
+                .addAll(symbolicLinkMappings)
+                .build();
+            return new DefaultSymbolicLinkMappings(newMappings, getRemappedSegments(currentRelativePathTracker.getSegments()));
+        }
+
+        @Override
+        public Iterable<String> getRemappedSegments(Iterable<String> segments) {
+            return Iterables.concat(prefixRelativePath, segments);
+        }
+    }
```

Comment:
I wonder if we could simplify this significantly. I think we only need a single type, something like:


```java
private static class SymbolicLinkMapping {

    public String remap(Path path) {
        return path.toString();
    }

    public SymbolicLinkMapping addMapping(String source, String target, RelativePathTracker currentRelativePathTracker) {
        SymbolicLinkMapping parent = this;
        return new SymbolicMapping() {
            public String remap(Path path) {
                String parentPath = parent.remap(path);
                return // actually remap path
            }
        }
    }
}
```

Reply:
I simplified it. There is now only one class, and we also don't need to loop at all.